File: Hybrid/src/main/java/siminov/hybrid/model/ApplicationDescriptor.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.hybrid.model;

import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;




/**
 * Exposes methods to GET and SET Application Descriptor information as per define in ApplicationDescriptor.si.xml file by application.
	<p>
		<pre>
		
Example:
	{@code

	<siminov>
	    
		<!-- General Application Description Properties -->
		
			<!-- Mandatory Field -->
		<property name="name">application_name</property>	
		
			<!-- Optional Field -->
		<property name="description">application_description</property>
		
			<!-- Mandatory Field (Default is 0.0) -->
		<property name="version">application_version</property>
	
	
		
		<!-- Siminov Framework Performance Properties -->
		
			<!-- Optional Field (Default is true)-->
		<property name="load_initially">true/false</property>
	
	
		
		<!-- Database Descriptors Used By Application (zero-to-many) -->	
			<!-- Optional Field's -->
		<database-descriptors>
			<database-descriptor>full_path_of_database_descriptor_file</database-descriptor>
		</database-descriptors>
			
	
	   	<!-- Services -->
	    <service-descriptors>
	  		
	  			<!-- Service -->
	        <service-descriptor>full_path_of_service</service-descriptor>
	    
	    </service-descriptors>
	
	    
		
	    <!-- Sync Handlers -->
	    	<!-- Sync Handler -->
	    <sync-descriptors>
	        
	        <sync-descriptor>
	            
	           		<!-- Mandatory Field -->
	     		<property name="name">name_of_sync_handler</property>
				
	     			<!-- Optional Field -->
	     		<property name="sync_interval">sync_interval_in_millisecond</property>
	     				
	     			<!-- Optional Field -->
	     				<!-- Default: SCREEN -->
				<property name="type">INTERVAL|SCREEN</property>
				
	     		<!-- Services -->
	     			<!-- Service -->
	     		<services>
	     		    
	     		    <service>name_of_service.name_of_api</service>
	     		    
	     		</services>
	        </sync-descriptor>
	        
	    </sync-descriptors>
	    
	
	    <!-- Push Notification -->
	    <notification-descriptor>
	        
	        	<!-- Optional Field -->
	        <property name="name_of_property">value_of_property</property>
	
	    </notification-descriptor>
	    
		
		<adapter-descriptors>
	        
	        <!-- Adapter Paths -->
		    <adapter-descriptor>adapter_path</adapter-descriptor>
		    	        
	    </adapter-descriptors>
			
		
		<!-- Library Descriptors Used By Application (zero-to-many) -->
			<!-- Optional Field's -->
		<library-descriptors>
		 	<library-descriptor>full_path_of_library_descriptor_file</library-descriptor>   
		</library-descriptors>
		
		
		<!-- Event Handlers Implemented By Application (zero-to-many) -->
		
			<!-- Optional Field's -->
		<event-handlers>
			<event-handler>full_java_class_path_of_event_handler/javascript_class_path_of_event_handler (ISiminovHandler/IDatabaseHandler)</event-handler>
		</event-handlers>
	
	</siminov>

	}
	
		</pre>
	</p>
 *
 */
public class ApplicationDescriptor extends siminov.connect.model.ApplicationDescriptor {

	private Collection<String> adapterDescriptorPaths = new ConcurrentLinkedQueue<String> ();
	
	private Map<String, AdapterDescriptor> adapterDescriptorsBasedOnName = new HashMap<String, AdapterDescriptor>();
	private Map<String, AdapterDescriptor> adapterDescriptorsBasedOnPath = new HashMap<String, AdapterDescriptor>();

	/**
	 * Get all adapter descriptors
	 * @return Adapter Descriptors
	 */
	public Iterator<AdapterDescriptor> getAdapterDescriptors() {
		return this.adapterDescriptorsBasedOnName.values().iterator();
	}
	
	/**
	 * Get adapter descriptor based on name
	 * @param adapterDescriptorName Name of adapter descriptor
	 * @return Adapter Descriptor
	 */
	public AdapterDescriptor getAdapterDescriptorBasedOnName(final String adapterDescriptorName) {
		return this.adapterDescriptorsBasedOnName.get(adapterDescriptorName);
	}

	/**
	 * Get adapter descriptor based on path
	 * @param adapterDescriptorPath Path of adapter descriptor
	 * @return Adapter Descriptor
	 */
	public AdapterDescriptor getAdapterDescriptorBasedOnPath(final String adapterDescriptorPath) {
		return this.adapterDescriptorsBasedOnPath.get(adapterDescriptorPath);
	}
	
	/**
	 * Add adapter descriptor
	 * @param adapterDescriptor Adapter Descriptor
	 */
	public void addAdapterDescriptor(final AdapterDescriptor adapterDescriptor) {
		this.adapterDescriptorsBasedOnName.put(adapterDescriptor.getName(), adapterDescriptor);
	}

	/**
	 * Add adapter descriptor based on its path
	 * @param adapterDescriptorPath Path of adapter descriptor
	 * @param adapterDescriptor Adapter Descriptor
	 */
	public void addAdapterDescriptor(final String adapterDescriptorPath, final AdapterDescriptor adapterDescriptor) {
		this.adapterDescriptorsBasedOnPath.put(adapterDescriptorPath, adapterDescriptor);
		this.adapterDescriptorsBasedOnName.put(adapterDescriptor.getName(), adapterDescriptor);
	}
	
	/**
	 * Check whether it contains adapter descriptor or not based on path
	 * @param adapterDescriptorPath Path of adapter descriptor
	 * @return (true/false) TRUE: If adapter descriptor exists | FALSE: If adapter descriptor does not exists.
	 */
	public boolean containAdapterDescriptorBasedOnPath(final String adapterDescriptorPath) {
		return this.adapterDescriptorsBasedOnPath.containsKey(adapterDescriptorPath);
	}
	
	/**
	 * Check if adapter descriptor exists based on its name
	 * @param adapterDescriptorName Name of adapter descriptor
	 * @return (true/false) TRUE: If adapter descriptor exists | FALSE: If adapter descriptor does not exists
	 */
	public boolean containAdapterDescriptorBasedOnName(final String adapterDescriptorName) {
		return this.adapterDescriptorsBasedOnName.containsKey(adapterDescriptorName);
	}
	
	/**
	 * Get all adapter descriptor paths
	 * @return Adapter Descriptor Paths
	 */
	public Iterator<String> getAdapterDescriptorPaths() {
		return this.adapterDescriptorPaths.iterator();
	}
	
	/**
	 * Add adapter descriptor path
	 * @param adapterDescriptorPath Path of adapter descriptor
	 */
	public void addAdapterDescriptorPath(final String adapterDescriptorPath) {
		this.adapterDescriptorPaths.add(adapterDescriptorPath);
	}
	
	/**
	 * Check if adapter descriptor path exists or not
	 * @param adapterDescriptorPath Path of adapter descriptor
	 * @return (true/false) TRUE: If path exists | FALSE: If path does not exists
	 */
	public boolean containAdapterDescriptorPath(final String adapterDescriptorPath) {
		return this.adapterDescriptorsBasedOnPath.containsKey(adapterDescriptorPath);
	}
	
	/**
	 * Remove adapter descriptor based on its name
	 * @param adapterDescriptorName Name of adapter descriptor 
	 */
	public void removeAdapterDescriptorBasedOnName(final String adapterDescriptorName) {
		
		Iterator<String> adapterDescriptorPaths = this.adapterDescriptorsBasedOnPath.keySet().iterator();
		
		String keyMatched = null;
		boolean found = false;
		while(adapterDescriptorPaths.hasNext()) {
			String adapterPath = adapterDescriptorPaths.next();
			
			AdapterDescriptor adapterDescriptor = this.adapterDescriptorsBasedOnPath.get(adapterPath);
			if(adapterDescriptor.getName().equalsIgnoreCase(adapterDescriptorName)) {
				keyMatched = adapterPath;
				found = true;
				break;
			}
		}
		
		if(found) {
			removeAdapterDescriptorBasedOnPath(keyMatched);
		} else {
			this.adapterDescriptorsBasedOnName.remove(adapterDescriptorName);
		}
		
	}
	
	/**
	 * Remove adapter descriptor based on path
	 * @param adapterDescriptorPath Path of adapter descriptor
	 */
	public void removeAdapterDescriptorBasedOnPath(final String adapterDescriptorPath) {
		AdapterDescriptor adapterDescriptor = this.adapterDescriptorsBasedOnPath.get(adapterDescriptorPath);
		
		this.adapterDescriptorsBasedOnName.remove(adapterDescriptor.getName());
		this.adapterDescriptorsBasedOnPath.remove(adapterDescriptorPath);
		this.adapterDescriptorPaths.remove(adapterDescriptorPaths);
	}
	
	/**
	 * Remove adapter descriptor 
	 * @param adapterDescriptor Adapter Descriptor
	 */
	public void removeAdapter(final AdapterDescriptor adapterDescriptor) {
		removeAdapterDescriptorBasedOnName(adapterDescriptor.getName());
	}
}

File: JavaScript/Siminov/Model/ApplicationDescriptor.js
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/



/**
	It contain all Models as per required by Siminov Framework.

	@module Model
*/
var win;
var dom;

try {

    if(!window) {
    	window = global || window;
    }

	win = window;
	dom = window['document'];
} catch(e) {
	win = Ti.App.Properties;
}



if(dom == undefined) {
    var Constants = require('../Constants');
    var Dictionary = require('../Collection/Dictionary');
    
    module.exports = ApplicationDescriptor;
    win.ApplicationDescriptor = ApplicationDescriptor;
}

/**
 	Exposes methods to GET and SET Application Descriptor information as per define in ApplicationDescriptor.si.xml file by application.
		
	Example:
		<siminov>
		
			<property name="name">SIMINOV HYBRID SAMPLE</property>	
			<property name="description">Siminov Hybrid Sample Application</property>
			<property name="version">0.9</property>
		
			<property name="load_initially">true</property>
		
			<!-- DATABASE-DESCRIPTORS -->
			<database-descriptors>
				<database-descriptor>DatabaseDescriptor.si.xml</database-descriptor>
			</database-descriptors>
		
			<!-- SIMINOV EVENTS -->
		 	<event-handlers>
			    <event-handler>siminov.hybrid.sample.events.SiminovEventHandler</event-handler>
			    <event-handler>siminov.hybrid.sample.events.DatabaseEventHandler</event-handler>
			</event-handlers>
				
		</siminov>


	@module Model	
	@class ApplicationDescriptor
	@constructor

 */
function ApplicationDescriptor() {

    var properties = new Dictionary();

	var databaseDescriptorPaths = new Array();
    var events = new Array();

	/**
	 	Get Application Descriptor Name as per defined in ApplicationDescriptor.si.xml file.
	 	
	 	@method getName
	 	@return {String} Application Descriptor Name.
	 */
	this.getName = function() {
        return properties.get(Constants.APPLICATION_DESCRIPTOR_NAME);
    }

	/**
	 	Set Application Descriptor Name as per defined in ApplicationDescriptor.si.xml file.
	 	
	 	@method setName
		@param name {String} Name of Application Descriptor.
	 */
	this.setName = function(name) {
        properties.add(Constants.APPLICATION_DESCRIPTOR_NAME, name);
    }

	/**
	 	Set Description of Application as per defined in ApplicationDescriptor.si.xml file.
	 
	 	@method getDescription
		@return {String} Description of application.
	 */
    this.getDescription = function() {
        return properties.get(Constants.APPLICATION_DESCRIPTOR_DESCRIPTION);
    }

	/**
	 	Set Description of Application as per defined in ApplicationDescriptor.si.xml file.
	 
	 	@method setDescription
		@param description {String} Description of application.
	 */
    this.setDescription = function(description) {
        properties.add(Constants.APPLICATION_DESCRIPTOR_DESCRIPTION, description);
    }

	/**
	 	Get Version of Application as per defined in ApplicationDescriptor.si.xml file.
		
		@method getVersion
		@return {String} Version of application.
	 */
    this.getVersion = function() {
        return properties.get(Constants.APPLICATION_DESCRIPTOR_VERSION);
    }

	/**
	 	Set Version of Application as per defined in ApplicationDescriptor.si.xml file.
	 
	 	@method setVersion
	 	@param version {String} Version of application.
	 */
    this.setVersion = function(version) {
        properties.add(Constants.APPLICATION_DESCRIPTOR_VERSION, version);
    }
    
    
    this.getDeploy = function() {
    	return properties.get(Constants.APPLICATION_DESCRIPTOR_DEPLOY);
    }
    
    this.setDeploy = function(deploy) {
    	properties.add(Constants.APPLICATION_DESCRIPTOR_DEPLOY, deploy);
    }
    

	/**
	 	Get all Properties defined in descriptor.
	 
	 	@method getProperties
		@return {Array} All Property Values.
	 */
	this.getProperties = function() {
		return properties.values();
	}


	/**
	 	Get Property based on name provided.
	 
	 	@method getProperty
		@param name {String} Name of Property.
		@return {String} Property value.
	 */
	this.getProperty = function(name) {
		return properties.get(name);
	}


	/**
	 	Check whether Property exist or not.
	 
	 	@method containProperty
		@param name {String} Name of Property.
	 	@return {Boolean} true/false, TRUE if property exist, FALSE if property does not exist.
	 */
	this.containProperty = function(name) {
		return properties.exists(name);
	}


	/**
	 	Add Property in property pool.
	 
	 	@method addProperty
		@param name {String} Name of Property.
	 	@param value {String} value of Property.
	 */
	this.addProperty = function(name, value) {
		properties.add(name, value);
	}
	

	/**
	 	Remove Property from property pool.
	 
	 	@method  removeProperty
		@param name {String} Name of Property.
	 */
	this.removeProperty = function(name) {
		properties.remove(name);
	}


	/**
	 	Add Database Descriptor path as per contained in ApplicationDescriptor.si.xml file.
	 
	 	@method addDatabaseDescriptorPath
	 	@param databaseDescriptorPath {String} DatabaseDescriptor path.
	 */
    this.addDatabaseDescriptorPath = function(databaseDescriptorPath) {
        databaseDescriptorPaths.push(databaseDescriptorPath);
    }

	/**
	 	Get all database descriptor paths as per contained in ApplicationDescriptor.si.xml file.
		
		@method getDatabaseDescriptorPaths
		@return {Array} It which contains all database descriptor paths.
	 */
    this.getDatabaseDescriptorPaths = function() {
        return databaseDescriptorPaths;
    }

	/**
		Get all event handlers as per defined in ApplicationDescriptor.si.xml file.
	 
 		@method getEvents
 		@return {Array} All event handlers defined in ApplicationDescriptor.si.xml file
	 */
    this.getEvents = function() {
        return events;
    }

	/**
	 	Add event as per defined in ApplicationDescriptor.si.xml file.
	 	
	 	@method addEvent
	 	@param event {String} Event Handler class name.
	 */
    this.addEvent = function(event) {
        events.push(event);
    }
}
